AB#703 elaborate description of remote attestation (#48)
edgelesssys/marblerun.sh: content/docs/features/attestation.md

# Attestation

> Hardware-rooted *remote attestation* is a key ingredient for distributed confidential apps. Marblerun relies heavily on the [*Data Center Attestation Primitives* (DCAP)](https://download.01.org/intel-sgx/sgx-dcap/1.9/linux/docs/Intel_SGX_DCAP_ECDSA_Orientation.pdf) of the latest SGX-enabled Intel Xeon processors.

Reviewer: Why cutting the introduction, too wordy?

Author: yes, I wanted to cap the total length of the article.

At the time of writing only Microsoft Azure had a public DCAP service deployed in their data-centers. Hence, our demos are heavily tested and deployed on Azure Kubernetes Service (AKS).
However, Marblerun works with any DCAP service complying with the SGX specification. You can read more about setting up your own DCAP infrastructure [in the Intel SGX development articles](https://software.intel.com/content/www/us/en/develop/articles/intel-software-guard-extensions-data-center-attestation-primitives-quick-install-guide.html).

## Coordinator deployment

> In a first step, the Marblerun Coordinator is deployed to the cluster.
> The Coordinator generates a X.509 certificate chain with a root and intermediate certificate authority (CA).
> It generates a remote attestation quote which contains the hash of the root certificate as well as a hardware measurement of the enclave the Coordinator is running in.

Reviewer: ```suggestion
Initially, the Marblerun Coordinator is deployed to the cluster.
The Coordinator generates an X.509 certificate chain with a root and intermediate certificate authority (CA).
It generates a remote attestation quote which contains the hash of the root certificate as well as a hardware measurement of its enclave.
```

> The admin revices the quote and verifies that measurement matches the Coordinator's binary to verify there was no manipulation to the Coordinators code.
> The admin also revices the certificate chain of the Coordinator and verifies that the hash of the root certificate is contained in the quote.
> This ensures that the certificate was indeed generated by the Coordinator and wasn't manipulated during transport.

Reviewer: ```suggestion
The admin receives the quote and verifies the Coordinator's integrity via the measurement. Next, they verify that the quote contains the root certificate’s hash. This ensures that the certificate was indeed generated by this Coordinator and was not manipulated during transport.
```

Author: Less linebreaks in the code?

Reviewer: Removing the linebreaks was unintentional, but I tried to shorten this section a bit.

After the successful verification of the Coordinator, the manifest can be uploaded to the cluster.

> The manifest describes which Marbles (services) should be deployed to your cluster. It contains a measurement values of the Marbles which were made after the Marbles were built.

Reviewer: ```suggestion
The manifest describes which Marbles (services) should be deployed to your cluster. It contains their enclaves’ build-time measurements. The manifest is described in more detail in our [“defining a manifest”] ]({{< ref " docs/workflows/define-manifest.md">}}) section. 
```

![Coordinator deployment](/img/coordinator_deployment.svg)

## Marble deployment and attestation

> The Coordinator enforces the manifest and deploys the desired Marble(s) to the cluster.
> Each Marble registers itself with a quote via gRPC at the coordinator.
> The quote and the manifest both contain measurement of the Marble, and the Coordinator compares both to ensure the Marble's identity and integrity.
> The [Marble's TLS credentials]({{< ref "docs/features/secrets-management.md" >}}) are generated by the coordinator and signed with the intermediate CA certificate.
> The Coordinator's certificate Chain was well as the freshly generated Marble certificate are then delivered to the Marble.

Reviewer: ```suggestion
This concludes the Coordinator’s initialization. From then on it enforces the manifest and only admits the desired Marble(s) to the cluster.
Each Marble registers itself with a quote via gRPC to the Coordinator.
The quote contains this Marble's measurements. The Coordinator compares them to the manifest to ensure the Marble's identity and integrity.
The [Marble's TLS credentials]({{< ref "docs/features/secrets-management.md" >}}) are generated by the Coordinator and signed with the intermediate CA certificate.
The Coordinator's certificate chain as well as the freshly generated Marble certificate are then delivered to the Marble.
```

![Marble deployment](/img/marble_deployment.svg)

## Confidential service mesh

> The Marble and the Coordinator can now communicate securely and authenticated over TLS (with client authentication), using the Coordinator's CA a root of trust.
> Different Marbles can communicate with each other in the same way.

Reviewer: ```suggestion
The Marble and the Coordinator can now communicate securely and authenticated over TLS (with client authentication), using the Coordinator's CA as a root of trust.
Different Marbles can communicate with each other in the same way.
```

> If a client (i.e. a relying party, a customer) wants to connect to a Marble, the Coordinators certificate is used to establish a TLS connection and verify the Marble's certificate.

Reviewer: ```suggestion
If a client (i.e., a relying party, a customer) wants to connect to a Marble, the Coordinator’s certificate is used as a trusted CA to establish a TLS connection and verify the Marble's certificate.
```

![service mesh](/img/service_mesh.svg)

## Cluster attestation

> The Coordinator issues one concise remote-attestation statement for your whole distributed app.
> A measurement of the coordinator is generated after the Coordinator's build and distributed to the relying party (this has to be done by the admin or operator).
> The relying party requests the Coordinator's quote and certificate chain.
> The quote contains the hash of the Coordinator's root CA certificate, with which the received certificate chain is verified.
> The quote also contains a measurement of the Coordinator, which is verified against the measurement received by the admin or operator.

Reviewer: ```suggestion
The Coordinator issues one concise remote-attestation statement for your whole distributed app.
The Coordinator's build-time measurement is distributed to the relying party (this must be done by the admin or operator).
The relying party requests the Coordinator's quote and certificate chain.
The quote contains the hash of the Coordinator's root CA certificate, which is verified against the received certificate chain.
The quote also contains the Coordinator's measurement, which is verified against the build-time measurement.
```

> The relying party then request the manifest from the Coordinator and ensures it contains the expected Marbles and the expected measurement values of those Marbles.
> The steps required on the client-side are described in our [verification hands-on]({{< ref "docs/workflows/verification.md" >}}).

Reviewer: ```suggestion
The relying party then requests the manifest from the Coordinator and ensures it contains the expected Marbles and their expected measurements.
The steps required on the client-side are described in our [verification hands-on]({{< ref "docs/workflows/verification.md" >}}).
```
